### PR TITLE
Feature 87 set answers without answer id to dummy answer

### LINKED ISSUE #87: Set answers without answerId to dummy answer

### PR DESCRIPTION
**Pull request by villejuutila**
Resolves #87

### CODE REVIEW

File: src/test/kotlin/fi/metatavu/oss/api/test/functional/tests/DeviceSurveyDataTestIT.kt [hunk 1]
```diff
@@ -1042,6 +1043,159 @@ class DeviceSurveyDataTestIT : AbstractResourceTest() {
         assertOffsetDateTimeEquals(OffsetDateTime.now(), OffsetDateTime.parse(answer2.metadata!!.createdAt))
     }
 
+    @Test
+    fun testSubmitMultiSelectAnswerWithoutAnswerOptionId() = createTestBuilder().use { testBuilder ->
```
Comment (anttileppa) [inline src/test/kotlin/fi/metatavu/oss/api/test/functional/tests/DeviceSurveyDataTestIT.kt:1047]: Maybe these tests should use mixed results where for example first 5 do not have an answer and 6-10 have

File: src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt [hunk 2]
```diff
@@ -177,6 +190,35 @@ class PageAnswerController {
         pageAnswerRepository.deleteSuspending(answer)
     }
 
+    /**
+     * Disconnects answer from device. To be used when devices are being removed in order to not lose
+     * the submitted answers
+     *
+     * @param it answer to disconnect
+     */
+    suspend fun unAssignFromDevice(it: PageAnswerBaseEntity) {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt:199]:
```suggestion
    suspend fun unassignFromDevice(it: PageAnswerBaseEntity) {
```

File: src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt [hunk 3]
```diff
@@ -331,4 +344,44 @@ class PageAnswerController {
...
 
+    /**
+     * Assigns a failsafe answer value to the answer in case it doesn't have one e.g. the answer is malformed
+     *
+     * @param pageQuestion question to get failsafe option for
+     * @param answer answer submitted from the device with malformed data
+     * @return answer with failsafe value
+     */
+    private suspend fun getFailsafeQuestionOption(pageQuestion: PageQuestionEntity, answer: DevicePageSurveyAnswer): DevicePageSurveyAnswer {
+        val dummyAnswer = questionOptionRepository.findFailsafeQuestionOption(pageQuestion.id) ?: createFailsafeQuestionOption(pageQuestion)
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt:355]:
```suggestion
        val failsafeAnswer = questionOptionRepository.findFailsafeQuestionOption(pageQuestion.id) ?: createFailsafeQuestionOption(pageQuestion)
```

File: src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt [hunk 4]
```diff
@@ -331,4 +344,44 @@ class PageAnswerController {
...
+
+        if (pageQuestion.type == FREETEXT) {
```
Comment (anttileppa) [inline src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt:357]: Shouldn't this be done before findFailsafeQuestionOption?

**Review by anttileppa — CHANGES_REQUESTED** [review]

File: src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt (hunk 3 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt:355]: Resolved

File: src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt (hunk 4 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt:357]: Resolved

File: src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt (hunk 2 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/oss/api/impl/pages/answers/PageAnswerController.kt:199]: Resolved

File: src/test/kotlin/fi/metatavu/oss/api/test/functional/tests/DeviceSurveyDataTestIT.kt (hunk 1 above)
Comment (villejuutila) [reply src/test/kotlin/fi/metatavu/oss/api/test/functional/tests/DeviceSurveyDataTestIT.kt:1047]: Resolved

**Review by anttileppa — APPROVED** [review]

**Pull request merged**